src/AutomationTests/AutomationIntegrationTests.cs

```diff
@@ -143,53 +145,118 @@ public void ScanAsync_WindowsFormsSampler_TaskIsCancelled_ThrowsCancellationExce
+            foreach (var result in results)
             {
-                return AsyncScanIntegrationCore(testAppPath, expectedErrorCount, expectedWindowCount);
+                ValidateOutput(result.WindowScanOutputs, WildlifeManagerKnownErrorCount);
             }
         }
 
-        private WindowScanOutput SyncScanIntegrationCore(string testAppPath, int expectedErrorCount, int expectedWindowCount = 1)
+        private WindowScanOutput ScanIntegrationCore(bool sync, string testAppPath, int expectedErrorCount, int expectedWindowCount = 1, int? processId = null)
         {
-            LaunchTestApp(testAppPath);
-            var builder = Config.Builder.ForProcessId(TestProcess.Id)
+            if (processId == null)
+            {
+                processId = LaunchTestApp(testAppPath);
+            }
+            var builder = Config.Builder.ForProcessId((int)processId)
                 .WithOutputDirectory(OutputDir)
                 .WithOutputFileFormat(OutputFileFormat.A11yTest);
 
             var config = builder.Build();
 
             var scanner = ScannerFactory.CreateScanner(config);
 
-            var output = ScanSyncWithProvisionForBuildAgents(scanner);
+            IReadOnlyCollection<WindowScanOutput> output;
+            if (sync)
+            {
+                output = ScanSyncWithProvisionForBuildAgents(scanner);
+            }
+            else
+            {
+                output = ScanAsyncWithProvisionForBuildAgents(scanner);
+            }
 
             return ValidateOutput(output, expectedErrorCount, expectedWindowCount);
         }
 
-        private WindowScanOutput AsyncScanIntegrationCore(string testAppPath, int expectedErrorCount, int expectedWindowCount = 1)
+        private IEnumerable<Task<ScanOutput>> GetAsyncScanTasks(string testAppPath, IEnumerable<CancellationToken> cancellationTokens)
         {
-            LaunchTestApp(testAppPath);
-            var builder = Config.Builder.ForProcessId(TestProcess.Id)
-                .WithOutputDirectory(OutputDir)
-                .WithOutputFileFormat(OutputFileFormat.A11yTest);
+            var taskFuncs = new List<Func<Task<ScanOutput>>>();
 
-            var config = builder.Build();
+            // Prepare scan tasks
+            foreach (var token in cancellationTokens)
+            {
+                var processId = LaunchTestApp(testAppPath);
+                var builder = Config.Builder.ForProcessId(processId)
+                    .WithOutputDirectory(OutputDir)
+                    .WithOutputFileFormat(OutputFileFormat.A11yTest);
 
-            var scanner = ScannerFactory.CreateScanner(config);
+                var config = builder.Build();
 
-            var output = ScanAsyncWithProvisionForBuildAgents(scanner);
+                var scanner = ScannerFactory.CreateScanner(config);
 
-            return ValidateOutput(output, expectedErrorCount, expectedWindowCount);
+                taskFuncs.Add(() => scanner.ScanAsync(null, token));
```

Comment:
Part of the goal here is to run the scans concurrently. On a fast machine, it might be possible for the first call to `scanner.ScanAsync` to complete before later calls to `scanner.ScanAsync` are even started. Is there a way that we can start all of the apps up front, possibly building their Scanners, then put the calls to `scanner.ScanAsync` into a tight loop after all of the potential delays are out of the way?

Reply:
That's what I'm attempting to do here- On this line (248) I'm creating a func instead of actually kicking off the scan. This way, we prepare everything (starting the app, creating the scanners, etc) ahead of time and then actually kick off the scans in a tight loop on line 255. Of course, there's the thread.sleep there to work around that UIAutomation issue for now, which means it's not really a tight loop at the moment, but that's a separate problem